secrets/secrets-vault/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+                    if let Err(err) = self.handle_get_request(reply_to.clone(), message).await {
+                        self.handle_get_request_error(reply_to, err).await;
+                    }
+                }
+                Ok(Operation::ServerXkey) => {
+                    self.handle_server_xkey_request(reply_to).await;
+                }
+                Err(err) => {
+                    self.handle_unsupported_request(reply_to, err).await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_unsupported_request(&self, reply_to: Subject, error: anyhow::Error) {
+        // TODO: handle the potential publish error
+        let _ = self
+            .nats_client
+            .publish(reply_to, error.to_string().into())
+            .await;
+    }
+
+    async fn handle_get_request_error(&self, reply_to: Subject, error: GetSecretError) {
+        // TODO: handle the potential publish error
+        let _ = self
+            .nats_client
+            .publish(reply_to, SecretResponse::from(error).into())
+            .await;
+    }
+
+    async fn handle_get_request(
+        &self,
+        reply_to: Subject,
+        message: Message,
+    ) -> StdResult<(), GetSecretError> {
+        if message.payload.is_empty() {
+            return Err(GetSecretError::Other("missing payload".to_string()));
+        }
+
+        let host_xkey = Self::extract_host_xkey(&message)?;
+
+        let secret_request = Self::extract_secret_request(&message, &self.xkey, &host_xkey)?;
+
+        let request_claims = Self::validate_and_extract_claims(&secret_request)?;
+
+        let policy = Self::validate_and_extract_policy(&secret_request)?;
+
+        let auth_claims = VaultAuthClaims {
+            iss: self.nkey.public_key(),
+            aud: self.vault_config.audience.clone(),
+            sub: request_claims.entity_id(),
+            exp: get_current_timestamp() + 60,
+            application: secret_request.context.application.name.clone(),
+            host: request_claims.host,
+            component: request_claims.component,
+            provider: request_claims.provider,
+        };
+
+        let encoding_key = Self::convert_nkey_to_encoding_key(&self.nkey)?;
+
+        let auth_jwt = Self::encode_claims_to_jwt(auth_claims, &encoding_key)?;
+
+        let vault_client =
+            Self::authenticate_with_vault(&self.vault_config, &policy, &auth_jwt).await?;
+
+        let secret_ref = VaultSecretRef::try_from(&secret_request)
+            .map_err(|e| GetSecretError::Other(e.to_string()))?;
+
+        let secret = Self::fetch_secret(&vault_client, &policy, secret_ref).await?;
+
+        let secret_response = SecretResponse {
+            secret: Some(Secret {
+                name: secret_request.name,
+                version: secret.metadata.version.to_string(),
+                string_secret: Some(secret.data.to_string()),
+                binary_secret: None,
```

Comment:
Ignorant question: are secrets in vault always strings?

Reply:
Great question, I can't speak for all of the backends but I believe that [Kv2](https://developer.hashicorp.com/vault/docs/secrets/kv/kv-v2), which is currently the only secrets engine this backend supports, generally expects you to store things as strings.